File: buildcontrol/app/sync/crypto.py
"""Optional payload encryption for public message brokers.

A free public MQTT broker is readable by anyone who guesses the topic, so the
change payloads can be sealed with a shared passphrase before they leave the
device. AES-256-GCM with a PBKDF2 key derived from the workspace key as salt —
every installation of the same workspace derives the same key from the same
passphrase, no key exchange required.
"""

from __future__ import annotations

import base64
import json
import os

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

MARKER = "a256gcm"
ITERATIONS = 200_000
NONCE_BYTES = 12


class DecryptionError(Exception):
    """Raised when a payload cannot be opened with the configured passphrase."""


def derive_key(passphrase: str, salt: str) -> bytes:
    """Derive the 32-byte content key from a passphrase and the workspace key."""
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt.encode("utf-8"),
        iterations=ITERATIONS,
    )
    return kdf.derive(passphrase.encode("utf-8"))


def encrypt(body: dict, key: bytes) -> bytes:
    """Seal a change envelope; returns the bytes to publish."""
    nonce = os.urandom(NONCE_BYTES)
    plaintext = json.dumps(body, ensure_ascii=False).encode("utf-8")
    ciphertext = AESGCM(key).encrypt(nonce, plaintext, None)
    envelope = {
        "enc": MARKER,
        "n": base64.b64encode(nonce).decode("ascii"),
        "c": base64.b64encode(ciphertext).decode("ascii"),
    }
    return json.dumps(envelope).encode("utf-8")


def decrypt(raw: bytes, key: bytes | None) -> dict:
    """Open a payload, transparently passing through unencrypted ones."""
    body = json.loads(raw.decode("utf-8"))
    if not isinstance(body, dict) or body.get("enc") != MARKER:
        return body
    if key is None:
        raise DecryptionError("payload is encrypted but no passphrase is configured")
    try:
        nonce = base64.b64decode(body["n"])
        ciphertext = base64.b64decode(body["c"])
        plaintext = AESGCM(key).decrypt(nonce, ciphertext, None)
    except Exception as exc:  # invalid key, tampering, truncation
        raise DecryptionError("wrong passphrase or damaged payload") from exc
    return json.loads(plaintext.decode("utf-8"))

File: buildcontrol_mobile/app/src/main/java/uz/buildcontrol/mobile/data/sync/Crypto.kt
package uz.buildcontrol.mobile.data.sync

import android.util.Base64
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec
import java.security.SecureRandom

/**
 * Optional payload encryption, byte-compatible with `app/sync/crypto.py`.
 *
 * A free public MQTT broker is readable by anyone who guesses the topic, so the
 * change envelopes are sealed with AES-256-GCM. The key is derived from the
 * shared passphrase with the workspace key as salt, so every installation of
 * the same workspace ends up with the same key without exchanging anything.
 */
object Crypto {

    const val MARKER = "a256gcm"
    private const val ITERATIONS = 200_000
    private const val KEY_BITS = 256
    private const val NONCE_BYTES = 12
    private const val TAG_BITS = 128

    class DecryptionException(message: String) : Exception(message)

    fun deriveKey(passphrase: String, salt: String): ByteArray {
        val spec = PBEKeySpec(passphrase.toCharArray(), salt.toByteArray(), ITERATIONS, KEY_BITS)
        return SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).encoded
    }

    fun encrypt(body: JsonObject, key: ByteArray): ByteArray {
        val nonce = ByteArray(NONCE_BYTES).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding").apply {
            init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(TAG_BITS, nonce))
        }
        val plaintext = SyncJson.encodeToString(JsonObject.serializer(), body).toByteArray()
        val ciphertext = cipher.doFinal(plaintext)
        val envelope = buildString {
            append("{\"enc\":\"").append(MARKER)
            append("\",\"n\":\"").append(b64(nonce))
            append("\",\"c\":\"").append(b64(ciphertext)).append("\"}")
        }
        return envelope.toByteArray()
    }

    /** Opens a payload, passing unencrypted ones straight through. */
    fun decrypt(raw: ByteArray, key: ByteArray?): JsonObject {
        val body = SyncJson.parseToJsonElement(raw.decodeToString()).jsonObject
        if (body["enc"]?.jsonPrimitive?.contentOrNull != MARKER) return body
        if (key == null) {
            throw DecryptionException("payload is encrypted but no passphrase is configured")
        }
        return try {
            val nonce = unb64(body.getValue("n").jsonPrimitive.content)
            val ciphertext = unb64(body.getValue("c").jsonPrimitive.content)
            val cipher = Cipher.getInstance("AES/GCM/NoPadding").apply {
                init(
                    Cipher.DECRYPT_MODE,
                    SecretKeySpec(key, "AES"),
                    GCMParameterSpec(TAG_BITS, nonce),
                )
            }
            SyncJson.parseToJsonElement(cipher.doFinal(ciphertext).decodeToString()).jsonObject
        } catch (exc: Exception) {
            throw DecryptionException("wrong passphrase or damaged payload")
        }
    }

    private fun b64(data: ByteArray): String = Base64.encodeToString(data, Base64.NO_WRAP)

    private fun unb64(text: String): ByteArray = Base64.decode(text, Base64.DEFAULT)
}
